zsh-window-title.zsh

```diff
@@ -4,139 +4,182 @@
+    add-zsh-hook preexec __zsh-window-title:preexec
+    add-zsh-hook preexec __zsh-window-title:preexec-update  # New hook for updating title before execution
+
+    # update window title when changing directories
+    add-zsh-hook chpwd __zsh-window-title:chpwd-update  # New hook for updating title on directory change
 }
 
 __zsh-window-title:init() {
-	'builtin' 'emulate' -LR zsh
+    'builtin' 'emulate' -LR zsh
 
-	'builtin' 'typeset' -gi ZWT_DEBUG=${ZWT_DEBUG:-$__zwt_debug_default}
+    'builtin' 'typeset' -gi ZWT_DEBUG=${ZWT_DEBUG:-$__zwt_debug_default}
 
-	'builtin' 'typeset' -gi ZSH_WINDOW_TITLE_DEBUG=${ZSH_WINDOW_TITLE_DEBUG:-$__zsh_window_title_debug_default}
+    'builtin' 'typeset' -gi ZSH_WINDOW_TITLE_DEBUG=${ZSH_WINDOW_TITLE_DEBUG:-$__zsh_window_title_debug_default}
 
-	__zsh-window-title:debugger
+    __zsh-window-title:debugger
 
-	'builtin' 'typeset' -gi ZSH_WINDOW_TITLE_DIRECTORY_DEPTH=${ZSH_WINDOW_TITLE_DIRECTORY_DEPTH:-$__zsh_window_title_directory_depth_default}
+    'builtin' 'typeset' -gi ZSH_WINDOW_TITLE_DIRECTORY_DEPTH=${ZSH_WINDOW_TITLE_DIRECTORY_DEPTH:-$__zsh_window_title_directory_depth_default}
 
-	__zsh-window-title:precmd
+    __zsh-window-title:precmd
 
-	'builtin' 'autoload' -U add-zsh-hook
+    'builtin' 'autoload' -U add-zsh-hook
 
-	__zsh-window-title:add-hooks
+    __zsh-window-title:add-hooks
 }
 
 __zsh-window-title:precmd() {
-	'builtin' 'emulate' -LR zsh
-	__zsh-window-title:debugger
-
-	local title=$(print -P "%$ZSH_WINDOW_TITLE_DIRECTORY_DEPTH~")
+    'builtin' 'emulate' -LR zsh
+    __zsh-window-title:debugger
 
-	'builtin' 'echo' -ne "\033]0;$title\007"
+    local current_dir=$(print -P "%$ZSH_WINDOW_TITLE_DIRECTORY_DEPTH~")  # Truncated current directory
+    __zsh-window-title:update "$current_dir"
 }
 
 __zsh-window-title:preexec() {
-	'builtin' 'emulate' -LR zsh
-	__zsh-window-title:debugger
+    'builtin' 'emulate' -LR zsh
+    __zsh-window-title:debugger
+
+    local cmd_first_part=${1[(w)1]}  # Store the first word of the command
+    local current_dir=$(print -P "%$ZSH_WINDOW_TITLE_DIRECTORY_DEPTH~")  # Truncated current directory
+    __zsh-window-title:update "$current_dir" "$cmd_first_part"
+}
+
+__zsh-window-title:update() {
+    local title_content="$1"
+    if [[ -n "$2" ]]; then
+        title_content="$title_content $2"
+    fi
+
+    # Set terminal title (xterm compatible)
+    printf "\033]0;%s\007" "$title_content"
+    printf "\033]0;%s\033\\" "$title_content"
 
-	local title=$(print -P "%$ZSH_WINDOW_TITLE_DIRECTORY_DEPTH~ - ${1[(w)1]}")
+    # Set tmux status bar title (xterm compatible)
+    printf "\033]2;%s\007" "$title_content"
+    printf "\033]2;%s\033\\" "$title_content"
+
+    # Set screen window name (specific to GNU Screen)
+    printf "\033k%s\033\007" "$title_content"
+    printf "\033k%s\033\\" "$title_content"
+
+    # Rename tmux window
+    if [ -n "$TMUX" ]; then
+        tmux rename-window "$title_content"
+    fi
```

Comment:
Can you explain what these are doing, and how I can test them?

Reply:
Added some comments to explain: 
```
    # Update title for xterm-compatible terminals
    'builtin' 'print' -Pn "\033]2;$title_content\033\\"
    # Update title for screen sessions
    'builtin' 'print' -Pn "\033k$title_content\033\\"
    # Update title for xterm-compatible terminals
    'builtin' 'print' -Pn "\033]0;$title_content\033\\"
    
    # Update title for tmux sessions if tmux is running
    if [ -n "$TMUX" ]; then
        'builtin' 'tmux' rename-window "$title_content"
    fi
```
    
    To test them Id recommend just pasting these into your terminal. I am trying to capture a wide array of compatability, support for tmux, screen, byobu, etc!
    
    They should rename the current tmux and screen panes and window with the current directory and command
    You can try to paste these in
```
print "\033]2;test\033\\"
print "\033ktest\033\\"
print "\033]0;test\033\\"
tmux rename-window "test"
```

Reply:
Added some comments to explain: 
    
You can test by pasting these commands into your terminal
```
    # Update title for xterm-compatible terminals
    print "\033]2;test1\033\\"
    # Update title for screen sessions
    print "\033ktest2\033\\"
    # Update title for xterm-compatible terminals
    print "\033]0;test3\033\\"
    'builtin' 'print' -Pn "\033]0;$title_content\033\\"
    
    # Update title for tmux sessions if tmux is running
    if [ -n "$TMUX" ]; then
        tmux rename-window "test4"
    fi
```

Reply:
Detailed Explanation:
print "\ek$title_content\e\\" sets the window title specifically for GNU screen sessions using the escape sequence \ek and \e\\.
print "\033]2;$title_content\033\\" updates the title of the terminal window for xterm-compatible terminals.
print "\033k$title_content\033\\" updates the tmux window name displayed in the tmux status bar.
print "\033]0;$title_content\033\\" updates both the terminal window title and the icon name for xterm-compatible terminals.
tmux rename-window "$title_content" renames the current tmux window to match the provided title content.

The difference between the \033]2; and \033]0; escape sequences lies in the specific window or terminal elements they target:

\033]2; (OSC 2)
This escape sequence is used to set the title of the terminal window itself.
It is commonly supported by xterm and xterm-compatible terminals.
When you use \033]2;$title_content\033\\, it sets the text displayed in the title bar of the terminal window.
\033]0; (OSC 0)
This escape sequence is used to set both the icon name and the title of the terminal window.
The icon name is a secondary title often used in window managers to represent the window in a minimized state or in task lists.
When you use \033]0;$title_content\033\\, it sets the text for both the terminal window title and the icon name.